dipy/reconst/dti.py

```diff
@@ -777,17 +782,35 @@ def fit(self, data, mask=None):
             min_signal = self.min_signal
 
         data_in_mask = np.maximum(data_in_mask, min_signal)
-        params_in_mask = self.fit_method(self.design_matrix, data_in_mask,
-                                         *self.args, **self.kwargs)
+
+        if self.return_S0_hat:
+            params_in_mask, model_S0 = self.fit_method(
+                self.design_matrix,
+                data_in_mask,
+                return_S0_hat=self.return_S0_hat,
+                *self.args,
+                **self.kwargs)
+        else:
+            params_in_mask = self.fit_method(self.design_matrix,
+                                             data_in_mask,
+                                             *self.args, **self.kwargs)
 
         if mask is None:
             out_shape = data.shape[:-1] + (-1, )
             dti_params = params_in_mask.reshape(out_shape)
+            if self.return_S0_hat:
```

Comment:
Could you explain how to simplify the if statements in the mask logic? I don't see how that would work. If `model_S0` is `None` then it would fail trying to reshape or it would return a large `None` array. 

I do agree that the ifs with `TensorFit` can be removed though to leave `return TensorFit(self, dti_params, model_S0=S0_params)` as long as `S0_params` is `None`.